tests/test_pipeline.py

```diff
@@ -0,0 +1,259 @@
+from typing import Any
+import concurrent.futures
+
+import pytest
+
+import psycopg
+from psycopg import pq
+from psycopg import errors as e
+
+pytestmark = pytest.mark.libpq(">= 14")
+
+
+def test_pipeline_status(conn: psycopg.Connection[Any]) -> None:
+    assert conn._pipeline is None
+    with conn.pipeline() as p:
+        assert conn._pipeline is p
+        assert p.status == pq.PipelineStatus.ON
+    assert p.status == pq.PipelineStatus.OFF
+    assert not conn._pipeline
+
+
+def test_pipeline_reenter(conn: psycopg.Connection[Any]) -> None:
+    with conn.pipeline() as p1:
+        with conn.pipeline() as p2:
+            assert p2 is p1
+            assert p1.status == pq.PipelineStatus.ON
+        assert p2 is p1
+        assert p2.status == pq.PipelineStatus.ON
+    assert conn._pipeline is None
+    assert p1.status == pq.PipelineStatus.OFF
+
+
+def test_cursor_stream(conn):
+    with conn.pipeline(), conn.cursor() as cur:
+        with pytest.raises(psycopg.ProgrammingError):
+            cur.stream("select 1").__next__()
+
+
+def test_server_cursor(conn):
+    with conn.cursor(name="pipeline") as cur, conn.pipeline():
+        with pytest.raises(psycopg.NotSupportedError):
+            cur.execute("select 1")
+
+
+def test_cannot_insert_multiple_commands(conn):
+    with pytest.raises(psycopg.errors.SyntaxError) as cm:
+        with conn.pipeline():
+            conn.execute("select 1; select 2")
+    assert cm.value.sqlstate == "42601"
+
+
+def test_pipeline_processed_at_exit(conn):
+    with conn.cursor() as cur:
+        with conn.pipeline() as p:
+            cur.execute("select 1")
+
+            # PQsendQuery[BEGIN], PQsendQuery
+            assert len(p.result_queue) == 2
+
+        assert cur.fetchone() == (1,)
+
+
+def test_pipeline_errors_processed_at_exit(conn):
+    conn.autocommit = True
+    with pytest.raises((e.OperationalError, e.UndefinedTable)):
```

Comment:
It doesn't seem that we need to catch `OperationalError` here, right? Or is it raised in some condition?
```suggestion
    with pytest.raises(e.UndefinedTable):
```